Python 3 imports: additional work on builtins, relative imports, importlib

josephsl opened the pull request:
### Link to issue number:
Fixes #8712
Fixes #8724
Fixes #8768

### Summary of the issue:
More work on Python 3 imports, including relative imports and changing __builtin__ to builtins, as well as using importlib to import modules at runtime.

### Description of how this pull request fixes the issue:
Changes made:

* Import __builtin__ -> builtins (Python Console, language handler).
* GUI and IAccessible objects: more relative imports (from . import ...).
* Use importlib to import modules at runtime, including for IAccessible modules.

### Testing performed:
Tested with Python 2.7 interpreter to make sure they are imported.

### Known issues with pull request:
None

### Change log entry:
None

josephsl committed:
Python 3/imports: additional relative imports (from . import module).

josephsl committed:
Python 3: more from . import module edits.

* IAccessible: from winWord import ... -> from .winWord import ...
* GUI: from addonGui import ... -> from .addonGui import ...

josephsl committed:
Python 3 import: __builtin__ -> builtins. Re #8724.

Python Console and language handler imports __builtin__, which has been renamed to builtins in Python 3.

josephsl committed:
Python 3/IAccessible objects: set import level to 1 when importing IAccessible support modules from the same folder. Re #8712.

Python 3 wants a dot (.) when importing modules from the same folder. When using __import__ function, the level argument specifies search level (default is 0). Thus allow other IAccessible modules to be loaded by specifying level of 1 (relative import from the same folder).

leonardder commented:
[**@josephsl**](https://github.com/josephsl) commented on [14 Sep 2018, 16:25 CEST](https://github.com/nvaccess/nvda/pull/8727#issuecomment-421375157 "2018-09-14T14:25:28Z - Replied by Github Reply Comments"):
> Hi, importlib is new in Python 3 (if I read this correctly).

It is also there in Python 2.7. 

> Help on built-in function __import__ in module __builtin__:
>
> __import__(...)
>     __import__(name, globals={}, locals={}, fromlist=[], level=-1) -> module
>    
>     Import a module. Because this function is meant for use by the Python
>    interpreter and not for general use, it is better to use
>    importlib.import_module() to programmatically import a module.

josephsl commented:
Hi, I see. In this case, I propose doing this via a new pull request. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Friday, September 14, 2018 12:11 PM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Author <author@noreply.github.com>
Subject: Re: [nvaccess/nvda] Python 3 imports: additional work on builtins, relative imports (#8727)

 

 <https://github.com/josephsl> @josephsl commented on 14 Sep 2018, 16:25 CEST <https://github.com/nvaccess/nvda/pull/8727#issuecomment-421375157> :

Hi, importlib is new in Python 3 (if I read this correctly).

It is also there in Python 2.7.

Help on built-in function import in module builtin:

import(...)
import(name, globals={}, locals={}, fromlist=[], level=-1) -> module

Import a module. Because this function is meant for use by the Python

interpreter and not for general use, it is better to use
importlib.import_module() to programmatically import a module.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8727#issuecomment-421456660> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkMnNPWG39Deuw8xp1sxP1qyVafteks5ua_81gaJpZM4WgJnV> .

leonardder commented:
I'd rather see this as part of this one. It creates unnecessary overhead if we have to change one line of code in this pr, and then change it again in the next.

See also #8768.

josephsl committed:
Merge branch 'master' into py3imports2

josephsl committed:
Module imports: use importlib.import_module to load modules dynamically at runtime. Re #8768.

Python 2.7 includes a base implementation of importlib which powers __import__ function, with Python 3 providing more features such as importers. Thus use importlib.import_module in app module handler and others, specifying appropriate package names in the process.

josephsl committed:
IAccessible: use importlib.import_module to load appropriate module. Re #8712.

Because importlib.import_module complains about module names, spell the module name (NVDAObjects.IAccessible.mod).

josephsl commented:
Hi,

By the way, there is an unused "import Queue" in synthDrivers.espeak (no underscore), so I'm ignoring that one. People who will be working on transition should note that this import causes an exception, so should be removed then.

Thanks.

josephsl committed:
App module handler/imports: typo.

josephsl committed:
Python 3/ctypes: HRESULT no longer exists in ctypes.wintypes.

HRESULT no longer exists in ctypes.wintypes in Python 3. The new location is ctypes.HRESULT. Even in Python 2, ctypes.wintypes.HRESULT points to ctypes.HRESULT.

leonardder commented on source/appModuleHandler.py:
```diff
@@ -93,8 +94,7 @@ def getAppNameFromProcessID(processID,includeExt=False):
 	# Try querying the app module for the name of the app being hosted.
 	try:
 		# Python 2.x can't properly handle unicode module names, so convert them.
-		mod = __import__("appModules.%s" % appName.encode("mbcs"),
```
This is definitely going to fail on python 3

```
>>> "hello, %s" % "test".encode("mbcs")
"hello, b'test'"
```

str.format gives the same result.

It should be sufficient to check if not isinstance(appName, str) before encoding to mbcs.

leonardder commented:
I'm again not very happy with the idea to touch code in one pr, knowing that it requires another change in a subsequent pr. then, we either better leave the code alone in this pr or fix it for good. I'm happy with either way, but not with leaving as is, honestly.

josephsl commented:
Hi, I’m willing to make an exception by addressing encoding problem as part of this PR with a comment stating that Python 3 improves Unicode support. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Monday, September 24, 2018 7:57 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Python 3 imports: additional work on builtins, relative imports, importlib (#8727)

 

I'm again not very happy with the idea to touch code in one pr, knowing that it requires another change in a subsequent pr. then, we either better leave the code alone in this pr or fix it for good. I'm happy with either way, but not with leaving as is, honestly.



—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8727#issuecomment-424004597> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AHgLkM0GjtdsQgRBupdEpav_TxDlS9Wiks5uePLVgaJpZM4WgJnV> .

josephsl committed:
Merge branch 'master' into py3imports2

josephsl committed:
Python 3/app module handler: only encode app name in MBCS if using Python 2, no longer needed in Python 3.

Addressing review comment (reviewed by Leonard de Ruijter/Babbage): appName.encode('mbcs') will fail in Python 3, so do not do it unless one is using Python 2.
Note that this is an exception to the PR stream: nowhere in import_module call does encoding takes place.

josephsl commented:
Hi,

Addressed the import concern as of today's commit. As noted in the commit message, I'm making an exception, as nobody else calling import_module calls name.encode apart from app module handler.

Thanks.

josephsl committed:
Python 3/app module handler: use sys.version_info.major.

leonardder commented:
Never mind, I was horribly wrong there. This place actually encodes (goes from unicode from bytes), my head was telling me that we had to decode (from bytes to unicode). So if it would have been isinstance, the right clause would be `if not isinstance(appname, str):`

josephsl commented:
Hi,

@feerrenrut, may I request a second look for this one? I'm interested in seeing if conflicts may arise when attempting to merge this pull request in the future.

Thanks.

feerrenrut reviewed (approved):
This looks good to me. Given that it is tested against python 2.7, I think we should merge it in now. We don't have any intention of continuing to support Python 2.7 once we move over to 3 so I will add a task to the python 3 project board to remove any 2.7 specific code.

leonardder commented:
I agree.

josephsl commented:
Possible regression reported: multi-byte file name issue. Multiple possibilities, including Python 2.7.16, this PR, or something else.

Thanks.

leonardder commented:
@josephsl: Could you please file this again against threshold?

josephsl commented:
Hi, yep that’ll be done (today at the earliest) with at least two separate PR’s: imports in one package, importlib in another. Thanks.

 

From: Leonard de Ruijter <notifications@github.com> 
Sent: Tuesday, June 4, 2019 12:29 AM
To: nvaccess/nvda <nvda@noreply.github.com>
Cc: Joseph Lee <joseph.lee22590@gmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [nvaccess/nvda] Python 3 imports: additional work on builtins, relative imports, importlib (#8727)

 

@josephsl <https://github.com/josephsl> : Could you please file this again against threshold?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub <https://github.com/nvaccess/nvda/pull/8727?email_source=notifications&email_token=AB4AXEE7F2SOX2QQ4RSJ7ETPYYKTJA5CNFSM4FUATHK2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODW3VVSY#issuecomment-498555595> , or mute the thread <https://github.com/notifications/unsubscribe-auth/AB4AXEGNIJMDUODN4Z4IMVDPYYKTJANCNFSM4FUATHKQ> .
